include/ensmallen_bits/sdp/lin_alg.hpp

```diff
@@ -33,43 +33,49 @@ inline size_t SvecIndex(size_t i, size_t j, size_t n)
  * @param input A symmetric matrix.
  * @param output Upper triangular representation.
  */
-inline void Svec(const arma::mat& input, arma::vec& output)
+template<typename MatAType, typename MatBType>
+inline void Svec(const MatAType& input, MatBType& output)
 {
-  const size_t n = input.n_rows;
+  MatBType iMat(input);
+  const size_t n = iMat.n_rows;
   const size_t n2bar = n * (n + 1) / 2;
 
-  output.zeros(n2bar);
+  output.zeros(n2bar, 1);
 
   size_t idx = 0;
   for (size_t i = 0; i < n; i++)
   {
     for (size_t j = i; j < n; j++)
     {
       if (i == j)
-        output(idx++) = input(i, j);
+        output(idx++, 0) = iMat(i, j);
       else
-        output(idx++) = arma::datum::sqrt2 * input(i, j);
+        output(idx++, 0) = arma::datum::sqrt2 * iMat(i, j);
     }
   }
 }
 
-inline void Svec(const arma::sp_mat& input, arma::sp_vec& output)
+template<typename MatAType, typename ElemType>
+inline void Svec(const MatAType& input,
+                 arma::SpMat<ElemType>& output)
 {
-  const size_t n = input.n_rows;
+  arma::SpMat<ElemType> iMat(input);
+
+  const size_t n = iMat.n_rows;
   const size_t n2bar = n * (n + 1) / 2;
 
   output.zeros(n2bar, 1);
 
-  for (auto it = input.begin(); it != input.end(); ++it)
+  for (auto it = iMat.begin(); it != iMat.end(); ++it)
```

Comment:
We don't know how well a given compiler will handle the `it != iMat.end();` part, ie. whether it will assume that `iMat.end();` is constant and hence not recreate the iterator for each iteration of the loop.  If it decides it can't assume that, we are wasting cycles here.

To avoid the above situation, I suggest declaring `auto it_end = iMat.end();` before the loop, and changing the loop declaration to:

    for (auto it = iMat.begin(); it != it_end; ++it)

Reply:
Agreed, thanks for pointing it out. :+1: